File: agri_farm/doc_events/task.py
import frappe,json
@frappe.whitelist()
def save_task(fr,date,doc,doc_name,end_date,r):

    doc=frappe.get_doc({
        "doctype":"Auto Repeat",
        "reference_doctype":doc,
        "reference_document":doc_name,
        "frequency":fr,
        "start_date":date,
        "end_date":end_date,
        "repeat_on_day":r
    })
    doc.insert()
    frappe.db.commit()

@frappe.whitelist()
def create_todo(responsible_person,task_name,status,priority,description, created_by=None,observer=None):
    count=0
    observers = json.loads(observer)
    f_observers = [i['user'] for i in observers]
    responsible_person = json.loads(responsible_person)
    assign_to = [i['user'] for i in responsible_person if not frappe.db.exists("ToDo", {'reference_name': task_name, 'allocated_to': i['user']})]
    print("ASIGN TOOO")
    print(json.dumps(assign_to))
    # data=frappe.db.sql(""" select name from `tabToDo` where reference_name=%s and allocated_to=%s and (status='Open' or status='Closed')""",(task_name,responsible_person),as_dict=1)
    # if data:

    # {
    #   'assign_to_me': '0',
    #   'assign_to': '["accounts@onlyorganic.co.in"]',
    #   'doctype': 'Task',
    #   'name': 'TASK-2022-00008',
    #   'bulk_assign': 'false',
    #   're_assign': 'false',
    #   'cmd': 'frappe.desk.form.assign_to.add'
    # }
    from frappe.desk.form.assign_to import add
    from frappe.share import add as share_add
    if len(assign_to) > 0:
        add({
            'assign_to_me': '0',
            'assign_to': json.dumps(assign_to),
            'doctype': 'Task',
            'name': task_name,
            'bulk_assign': 'false',
            're_assign': 'false',
        })
        if created_by:
            share_add("Task", task_name, created_by, write=1)
        if observer:
            for y in f_observers:
                share_add("Task", task_name, y, write=1)
        for x in assign_to:
            share_add("Task", task_name,x, write=1)

    todos = frappe.db.sql(""" SELECT * FROM `tabToDo` WHERE reference_name=%s """, task_name,as_dict=1)

    for t in todos:
        share_add("ToDo", t.name, t.allocated_to, write=1)
        # if observer:
        #     for y in f_observers:
        #         share_add("ToDo",  t.name, y)
        # for x in assign_to:
        #     share_add("ToDo",  t.name, x)
    # if frappe.db.exists("ToDo",{'reference_name':task_name,'allocated_to':responsible_person}):
    #     print('!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! already created')
    # else:
    #     if priority=='Urgent':
    #         priority='High'
    #     new_todo =frappe.get_doc({"doctype":"ToDo",
    #     "reference_type":'Task',
    #     "reference_name" : task_name,
    #     "allocated_to" : responsible_person,
    #     "status" : status,
    #     "priority" : priority,
    #     "description" : description,
    #     })
    #     new_todo.insert()
    #     frappe.db.commit()





#  ----------------------------------------------------------------------------------------------------------  
# @frappe.whitelist()
# def completed_validation(task_name):
#     data=frappe.db.sql(""" select name,status from `tabToDo` where reference_name=%s""",task_name,as_dict=1)
#     for i in data:
#         if i.status=='Open':
#             frappe.throw('ToDo {0} is not Completed'.format(i.name))


# @frappe.whitelist()
# def completed_val(doc,method):
#     print("hhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhh")
#     if doc.status =='Completed':
#         print('ffffff')
#         data=frappe.db.sql(""" select name,status from `tabToDo` where reference_name=%s""",doc.name,as_dict=1)
#         if data:
#             for i in data:
#                 if i.status=='Open':
#                     frappe.throw('ToDo {0} is not Completed'.format(i.name))
# ---------------------------------------------------------------------------------------------------------------
# def save_task(fr,date,doc,doc_name,end_date,r):
#     if fr or date: 
#         print("Taaaaaaaaaaaaaaaaaaaaaaaaadsfsdgdfhgfjhg")

#         doc=frappe.get_doc({"doctype":"Auto Repeat",
#             "reference_doctype":doc,
#             "reference_document":doc_name,
#             "frequency":fr,
#             "start_date":date,
#             "end_date":end_date,
#             "repeat_on_day":r

#         })
#         doc.insert()
#         frappe.db.commit()
@frappe.whitelist()
def fetch_company(pro_name):
    data=frappe.db.sql("""select company from `tabProject` where project_name=%s""",pro_name,as_dict=1)
    print("company")
    print(data)
    return data

File: agri_farm/public/js/task.js
// Copyright (c) 2015, Frappe Technologies Pvt. Ltd. and Contributors
// License: GNU General Public License v3. See license.txt


frappe.ui.form.on("Task", {
    company_name:function(frm){
        if(cur_frm.doc.company_name){
            cur_frm.set_value("company",cur_frm.doc.company_name)
            cur_frm.refresh_field("company")

        }

    },
    project:function(frm){
        if(cur_frm.doc.project){
            console.log("byeeee")
            console.log(cur_frm.doc.project_name)
            frappe.call({
                method:"agri_farm.doc_events.task.fetch_company",
                args:{
                    pro_name:cur_frm.doc.project_name
                },
                callback:function(r){
                    cur_frm.set_value("company_name",r.message[0].company)
                    console.log(r.message[0].company)
                    cur_frm.refresh_field("company_name")
                    cur_frm.set_value("company",r.message[0].company)
                    cur_frm.refresh_field("company")

                }
            })
        }

    },
	// onload:function(frm){
        
     //    // cur_frm.set_df_property("frequency","hidden",1)
     //    // cur_frm.set_df_property("start_date","hidden",1)
     //    // cur_frm.set_df_property("end_date","hidden",1)
     //    // cur_frm.set_df_property("repeat_on_day","hidden",1)
	// },
    onload_post_render(frm) {
            cur_frm.set_df_property("is_template","hidden",1)
            


        // if(!window.location.hash) {
        //     window.location = window.location + '#';
        //     window.location.reload();
        // }
        // intro(frm)
        set_status(frm)
    },
    // before_save:function(frm){
    //     if(cur_frm.doc.status=='Completed'){
    //         frappe.call({
    //             method:"agri_farm.doc_events.task.completed_validation",
    //             args:{
    //                 task_name:cur_frm.doc.name,  
    //             },
    //             callback:function(r){
    //                 console.log("function executed")
    //             }
    //         })

    //     }
    // },

    
   
    
    // repeat_task:function(frm){
    //     if(cur_frm.doc.repeat_task){
    //         frm.set_df_property("frequency","hidden",0)
    //         frm.set_df_property("start_date","hidden",0)
    //         frm.set_df_property("end_date","hidden",0)
    //         frm.set_df_property("repeat_on_day","hidden",0)
    //         frm.set_df_property("frequency","reqd",1)
    //         frm.set_df_property("start_date","reqd",1)
    //         frm.set_df_property("end_date","reqd",1)
    //         frm.set_df_property("repeat_on_day","reqd",1)
            
    //     }
    //     else{
    //         cur_frm.set_df_property("frequency","hidden",1)
    //         cur_frm.set_df_property("start_date","hidden",1)
    //         cur_frm.set_df_property("end_date","hidden",1)
    //         cur_frm.set_df_property("repeat_on_day","hidden",1)
    //         frm.set_df_property("frequency","reqd",0)
    //         frm.set_df_property("start_date","reqd",0)
    //         frm.set_df_property("end_date","reqd",0)
    //         frm.set_df_property("repeat_on_day","reqd",0)
    //     }
    // },
    after_save:function(frm){
        var rr=cur_frm.doc.repeat_task
        if(rr){
            frappe.call({
                method:"agri_farm.doc_events.task.save_task",
                args:{
                    fr:cur_frm.doc.frequency,
                    date:cur_frm.doc.start_date,
                    doc:cur_frm.doc.doctype,
                    doc_name:cur_frm.doc.name,
                    end_date:cur_frm.doc.end_date,
                    r:cur_frm.doc.repeat_on_day
                },
                async: false,
                callback:function(r){
                    console.log("function executed. AFTER SAVE. REPEAT TASK")
                }
            })
        }
       
        if(cur_frm.doc.status==='Open'){
            console.log("joooo")
            if(cur_frm.doc.responsible_person){
                var rp=frm.doc.responsible_person
                // for(var i=0;i<rp.length;i++){
                    frappe.call({
                        method:"agri_farm.doc_events.task.create_todo",
                        args:{
                            responsible_person:cur_frm.doc.responsible_person.concat(cur_frm.doc.participants),
                            task_name:cur_frm.doc.name,
                            status:cur_frm.doc.status,
                            priority:cur_frm.doc.priority,
                            description:cur_frm.doc.subject,
                            created_by: cur_frm.doc.created_by,
                            observer: cur_frm.doc.observer,
                        },
                        async: false,
                        callback:function(r){
                        set_status(frm)
                            set_status(frm)
                            cur_frm.reload_doc()
                            console.log("function executed. CREATE TODO")
                        }
                    })
            
                // }
            }
            // if(cur_frm.doc.participants){
            //     var p=frm.doc.participants
            //     // for(var i=0;i<p.length;i++){
            //     //     console.log(cur_frm.doc.participants[i].user)
            //         frappe.call({
            //             method:"agri_farm.doc_events.task.create_todo",
            //             args:{
            //                 responsible_person:cur_frm.doc.participants,
            //                 task_name:cur_frm.doc.name,
            //                 status:cur_frm.doc.status,
            //                 priority:cur_frm.doc.priority,
            //                 description:cur_frm.doc.subject,
            //
            //
            //             },
            //             async:false,
            //             callback:function(r){
            //                 // set_status(frm)
            //                 cur_frm.reload_doc()
            //                 console.log("function executed. PARTICIPANTS")
            //             }
            //         })
            //
            //     // }
            //
            // }
            
        }
        // cur_frm.reload_doc()
        // intro(frm)
        // set_status(frm)
    }

   


})
function set_status(frm){
   // if(!frm.is_new())
   //  {
   //      frm.set_intro(('{0}',
   //      [
   //          '<h3>ToDo for this Task</h3>'
   //          ]
   //      )
   //          )
   //      // return true
   //  }
    var params = {
        'task': frm.doc.name
    }
    frappe.call('agri_farm.task_api.get_todo', params).then((r) => {
        let total = r.message;
        var name = total.name;
        var status = total.status1;
        var reference_name = total.reference_name;
        var size = name.length;

        var todos = cur_frm.is_new() ? "":"<h3>ToDo for this Task</h3>"
        for(var i = 0; i< size; i++){
            todos += `<a href="/app/todo/${name[i]}">${i+1}: ${name[i]} -${reference_name[i]} </a>`
            todos += `<p>${status[i]}</p>`
        }
        frm.set_intro(('',todos), 'red');

    })
    
}